pandas/plotting/_matplotlib/hist.py

```diff
@@ -243,15 +244,27 @@ def _grouped_hist(
     sharey : bool, default False
     rot : int, default 90
     grid : bool, default True
+    legend: : bool, default False
     kwargs : dict, keyword arguments passed to matplotlib.Axes.hist
 
     Returns
     -------
     collection of Matplotlib Axes
     """
 
+    if legend and "label" not in kwargs:
+        if isinstance(data, ABCDataFrame):
+            if column is None:
+                kwargs["label"] = data.columns
+            else:
+                kwargs["label"] = column
+        else:
+            kwargs["label"] = data.name
```

Comment:
emm, I do not think the logic here is correct. We want group keys as labels, not column names, right? so, in your case in tests, labels should be `1` and `2` not `a` and `b`. And `a` and `b` should be the suptitles of subplots (and should already be plotted out).

Reply:
This section of code is used by the 'by' argument is specified. Very hard to tell from the values used in the test, so I've switched over to using something simple:

````
index = 3 * ["1"] + 3 * ["2"]
df = DataFrame([(3, 4)]*3 + [(5, 6)]*3, index=index, columns=["a", "b"])
df.index.names = ['c']
ret = df.hist(by='c', legend=True)
````

This produces

![image](https://user-images.githubusercontent.com/45562402/82497370-2136f380-9abc-11ea-912f-20e23fdf4c8c.png)

and similarly for series:

````
index = 3 * ["1"] + 3 * ["2"]
df = DataFrame([(3, 4)]*3 + [(5, 6)]*3, index=index, columns=["a", "b"])
df.index.names = ['c']
ret = df.a.hist(by='c', legend=True)
````

![image](https://user-images.githubusercontent.com/45562402/82497414-30b63c80-9abc-11ea-80c4-a55cb0a009eb.png)

It appears correct to me.

Reply:
okay, nice!